shared/src/main/scala-2.10/scalajson.ast/package.scala

```diff
@@ -343,4 +345,280 @@ package object ast {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        val maxLenCheck = digitLength <= maxLengthConstant
+
+        if (result < multmin && maxLenCheck && !eFlag)
+          return None
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            if (result < limit + digit && maxLenCheck)
+              return None
+            result -= digit
+          }
+        }
+      }
+      i += 1
+    }
+    if (result < limit)
+      None
+    else {
+      if (negative)
+        Some(result.toLong)
+      else
+        Some(-result.toLong)
+    }
+  }
+
+  private[ast] def toBigInt(value: String): Option[BigInt] = {
+    var decimalFlag = false
+    var result: BigInt = 0
+    var negative = false
+    var char: Char = 0
+    var i = 0
+    var eFlag = false
+    var trailingZeroes: BigInt = 0
+    var negativeEFlag: Boolean = false
+    var resultNegativeEFlag: BigInt = 0
+    var digitLength: BigInt = 0
+    if (value.charAt(0) == '-') {
+      negative = true
+      i += 1
+    }
+
+    while (i < value.length) {
+      char = value.charAt(i)
+      if (char == '.')
+        decimalFlag = true
+      else if ((char | 0x20) == 'e') {
+        eFlag = true
+        val charNext = value.charAt(i + 1)
+        if (charNext == '-')
+          negativeEFlag = true
+        if (negativeEFlag || charNext == '+') {
+          i += 1
+        }
+      } else {
+        if (!(eFlag || decimalFlag))
+          digitLength += 1
+
+        val digit = Character.digit(char, radix)
+
+        if (digit == 0)
+          if (!decimalFlag)
+            trailingZeroes += 1
+          else if (!decimalFlag)
+            trailingZeroes = 0
+
+        if (digit != 0 && (decimalFlag || eFlag))
+          if (!negativeEFlag)
+            return None
+          else {
+            if (trailingZeroes != 0) {
+              resultNegativeEFlag *= radix
+              resultNegativeEFlag += digit
+              if (trailingZeroes >= resultNegativeEFlag) {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            result -= digit
+          }
+        }
+      }
+      i += 1
+    }
+    if (negative)
+      Some(result)
+    else
+      Some(-result)
+  }
+
+  private[ast] def toDouble(value: String): Option[Double] = {
+    try {
+      val asDouble = value.toDouble
+      if (BigDecimal(value, MathContext.UNLIMITED) == BigDecimal(asDouble, MathContext.UNLIMITED))
```

Comment:
I don't know what "without losing any data" means, at least not with this implementation.

```scala
scala> BigDecimal.exact(0.1)
res11: scala.math.BigDecimal =
0.1000000000000000055511151231257827021181583404541015625
```

That's what the `Double` value _actually_ is for the thing we call `0.1` but then try to represent as IEEE 754 binary64.

I think there are three justifiable positions, depending on how much work the library is supposed to do
1. Double conversion never fails; you get the "closest" `Double` possible even if it might be 0 or +- infinity.  Plus side: super simple.  Minus side: clearly unrepresentable stuff gets through.
2. Double conversion fails when clearly wrong ("1e-1000" or "1e1000" for instance) but otherwise gives the best finite nonzero approximation (or zero if it's truly zero).  (TBD: policy for bad mismatches with subnormal numbers.)  Plus: helpful to avoid the worst errors, matches compiler treatment of constants (if 1e-1000 -> Some(0.0)).  Minus: have to think about errors, doesn't catch all of them.
3. Double conversion succeeds when the string is any valid rounding of a binary64 value encoded fully as decimals; otherwise it fails.  So 0.1, 0.100000000000000006, 0.1000000000000000056, etc. are all fine.  Plus: catches everything that is reasonably an error.  Minus: catches a lot of things that seem as reasonable as anything with Double does; very slow and lots of work to implement.

In general, `Double` _always_ means "might have lost some data".  Trying to save people from data loss in this particular case seems both difficult and almost-pointless.  If `0.3 - (0.1+0.2)` in `Double`s isn't `0`, and it's not, we have bigger problems than supposing that "0.11111111111111111" in JSON has a sensible `Double` representation.  (It fails the "valid rounding" test.)